### PR TITLE
Torch improvements

### PR DESCRIPTION
**Pull request by SergeyTsimfer**
This PR proposes to:

## Refactor the inputs/targets logic of `TorchModel`. 
Previously, we allowed to use `*args`, `**kwargs` and even `feed_dict` in both `train` and `predict` methods, as well as in the visualization tools like `get_gradcam`. This was ugly in code (and required ~50+ lines of parsing), and was prone to errors, especially in the multi-input or multi-output settings. It is very unintuitive to use as you can essentially pass anything to a model and hope it would arrange elements in correct order. Moreover, as `targets` have no defined argument, they are parsed by using hardcoded names, which is very confusing when works incorrectly.

In this PR, a different logic is implemented. All of the above methods now accept `inputs` and `targets` arguments, that are passed directly to the model or loss computation function.
Inside the code, they are always treated like sequences of `np.narrays`, which makes it easy to, for example, split both inputs and targets into microbatches, compute their shapes, and assert some additional properties for better user experience.
The only thing we (possibly) do with those arguments is, if there is only one element, we pass it to model / loss directly, so you don't need to worry about single-element sequences inside your `forward` pass.

## Improve the `fetches` behavior
The `fetches` parameter is renamed to `outputs`.
As previously, one can use pre-defined aliases to get parts of the network, i.e. `predictions` and `loss`. The list of aliases is extended by:
- `predictions_{i}` to get the i-th tensor of model output
- `predictions_{i}_{j}` to get the j-th operation, defined in the `outputs` config section, applied to the i-th tensor. For example, if the `config['outputs'] = {'first_tensor': ['sigmoid', 'softmax', custom_callable]}`, then the `predictions_0_1` gets the result of `softmax` operation, applied to the first tensor

Moreover, now we accept layer identifiers to get the intermediate activations from a neural network. The layer identifier is a string that defines how to get a layer in the model, and uses the same semantics, as the usual `getattr`/`getitem` way.

For example, if you can access a particular layer by 
```python
batchflow_model.model.body.encoder["block-1"][0].layer
```
then you can pass the 
```python
'model.body.encoder["block-1"][0].layer'
```
string as one of the `outputs` to retrieve the activations from exactly this part of the NN. Note that all of the layer identifiers start with `'model.'` prefix.

This logic is currently used only in `predict` method, not in `train`. Comment, if you see the use case for it to be in `train` as well.

## Re-arrange parts of the code to be more cohesive
The `TorchModel` code serves two main purposes: to create a `PyTorch model` instance and to provide infrastructure for its training. Now there is a clearer separation between the two.

Also, the `build` stage of the `TorchModel` instance is now explicitly split into `update instance attributes from passed config` and `update config with parsed attributes` stages. 

## Improve textual and graphical visualizations
- fixed a lot of newlines in the `.information` visualization
- added a loss plot
- more info is saved to `last_iteration_info` dictionary during the `train` pass
- a lot of methods are now public, so they have documentation and explanation of inner workings. Also, methods are reoranged to better correspond to logic of their usage

# Proposals
Additionally, I think that it would be beneficial to:

- Remove the `BaseModel` or at least move all the loading logic from it to the `TorchModel` itself. 
Now, as there is only one framework we support (#629), it would be easier to comprehend if all the code regarding `TorchModel` would be in the same place.

- Rename `.config` and `.full_config` attributes into `.external_config` and `.config` respectively. This would make it easier to understand the meaning of all of these configs.

- Make a separate `model_config`, which is subset of the main configuration by `order` and its elements keys.
As of now, there is no clear explanation to why we need to first parse instance attributes from config and then update config with some of the instance attributes. In reality, we need to update only the `model_config`, which makes sense, as model creation parameters can be unknown before we get additional information about the runtime (for example, we can infer shapes from passed batch data, which are unknown at instance initialization from config).

Update 15/12/2021: added `microbatch` to `predict`, as well as the `layer id` parsing logic to train.

### CODE REVIEW

File: batchflow/models/torch/base.py [hunk 1]
```diff
@@ -736,44 +653,116 @@ def make_decay(self, decay, optimizer=None, **kwargs):
...
+        The order is defined by `order` key in the config, which is [`initial_block`, `body`, `head`] by default.
+        Each item in `order` should describe the block name, the config name and method to create. It can be a:
+            - string, then we use at as name, key and method name
+            - tuple of three elements, which are name, key and method name or callable
+            - dictionary with three items, which are `block_name`, `config_name` and `method`.
+
+        The `block_name` is then used as the identifier in resulting model, i.e. `model.initial_block`, `model.body`.
+        The `config_name` is used to retrieve block creation parameters from config.
+        The `method` is either a string to getattr from the current instance or a callable.
+        Either method or callable should return an instance of nn.Module and accept block parameters.
```
Comment (a-arefina) [inline batchflow/models/torch/base.py:679]: indent this block to match the dictionary description above

Last line is not quite clear: what do you refer to by 'method'

Comment (a-arefina) [reply batchflow/models/torch/base.py:679]: "name of the method to get from the current instance" - this description from below is much more clear

File: batchflow/models/torch/base.py [hunk 2]
```diff
@@ -736,44 +653,116 @@ def make_decay(self, decay, optimizer=None, **kwargs):
...
+                block_name = config_name = method = item
+            elif isinstance(item, tuple) and len(item) == 3:
+                block_name, config_name, method = item
+            elif isinstance(item, dict):
+                block_name = item['block_name']
+                config_name = item.get('config_name', block_name)
+                method = item.get('method', config_name)
+
+            # Make block, from the `inputs`, transfer it to device
+            # Important: apply to the `inputs` before showing to the next block, so the shapes/etc are updated
```
Comment (a-arefina) [inline batchflow/models/torch/base.py:699]:
```suggestion
            # Important: apply to the `inputs` before passing them to the next block, so the shapes/etc are updated
```

File: batchflow/models/torch/base.py [hunk 3]
```diff
@@ -736,44 +653,116 @@ def make_decay(self, decay, optimizer=None, **kwargs):
...
+        block = config[name]
+
+        if isinstance(block, nn.Module):
+            # Already initialized module
+            pass
+
+        elif isinstance(block, dict):
+            block_params = {**config['common'], **block}
+
+            if 'module' in block_params:
```
Comment (a-arefina) [inline batchflow/models/torch/base.py:726]: we also have `base` and `base_block`

File: batchflow/models/torch/base.py [hunk 4]
```diff
@@ -863,63 +835,67 @@ def transfer_to_device(self, data):
...
-
-        output = []
-        for name in _fetches:
-            if name in outputs:
-                value = outputs[name]
-                value = self.transfer_from_device(value)
-                output.append(value)
-            else:
-                raise KeyError('Unknown value to fetch', name)
+        raise TypeError(f'Passed data should either be a `torch.Tensor` or a container of them, got {data}')
```
Comment (a-arefina) [inline batchflow/models/torch/base.py:856]: numpy arrays and sclalars are also allowed

File: batchflow/models/torch/base.py [hunk 5]
```diff
@@ -863,63 +835,67 @@ def transfer_to_device(self, data):
...
+            Desired outputs of the method.
+            Each string defines a tensor to get and should be one of pre-defined or set in `outputs` key in the config.
+            Pre-defined tensors are `predictions`, `loss`, and `predictions_{i}` for multi-output models.
         lock : bool
             If True, then model, loss and gradient update operations are locked, thus allowing for multithreading.
         sync_frequency : int, bool or None
             If int, then how often to apply accumulated gradients to the weights.
-            If True, then value from config is used (default value is to apply gradients after each batch of data).
+            If True, then value from config is used. Default value is 1,
+            which means to apply gradients after each batch of data.
```
Comment (a-arefina) [inline batchflow/models/torch/base.py:890]:
```suggestion
            If True, then value from config is used. 
            Default value is 1, which means to apply gradients after each batch of data.
```

File: batchflow/models/torch/base.py [hunk 6]
```diff
@@ -1200,174 +1210,128 @@ def predict(self, *args, targets=None, feed_dict=None, fetches=None, lock=True,
...
+            # For each operation, add multiple aliases
+            for j, operation in enumerate(output_operations):
+                output_tensor, operation_name = self.apply_output_operation(tensor, operation)
+                if operation_name:
+                    outputs[output_prefix + operation_name] = output_tensor # i.e. `first_sigmoid`, `sigmoid`
 
-            prefix = [*ops.keys()][i]
-            attr_prefix = prefix + '_' if prefix else ''
+                outputs.update({
+                    output_prefix + str(j) : output_tensor, # i.e. `first_0`, `0`
```
Comment (a-arefina) [inline batchflow/models/torch/base.py:1304]: so this key is overitten for multiple predictions?

File: batchflow/models/torch/visualization.py [hunk 7]
```diff
@@ -97,6 +109,17 @@ def save_graph(self, log_dir=None, **kwargs):
         writer.add_graph(self.model, self._placeholder_data())
         writer.close()
 
+    def show_loss(self):
```
Comment (a-arefina) [inline batchflow/models/torch/visualization.py:112]: add yscale=log option

**Comment by SergeyTsimfer** [comment]
Added `microbatch` to `predict`, as well as the `layer id` parsing logic to train.

File: batchflow/models/torch/base.py [hunk 8]
```diff
@@ -1167,94 +1151,195 @@ def _train_sam_update_gradients(self, inputs, targets, sync_frequency, sam_rho,
...
+        >>> batchflow_model.predict(inputs=B.images, outputs='model.body.encoder["block-0"]')
         """
         # Acquire lock; release in any case
         try:
             if lock:
                 self.model_lock.acquire()
+            self.last_predict_info = {}
+
+            # Parse inputs and targets: always a list
+            inputs = list(inputs) if isinstance(inputs, (tuple, list)) else [inputs]
```
Comment (a-arefina) [inline batchflow/models/torch/base.py:1199]: to_list?

https://github.com/analysiscenter/batchflow/blob/7d201bfdf220496e966e9db5eb27a57212638994/batchflow/utils.py#L19

File: batchflow/models/torch/base.py [hunk 9]
```diff
@@ -1167,94 +1151,195 @@ def _train_sam_update_gradients(self, inputs, targets, sync_frequency, sam_rho,
...
+
+    # Common utilities for train and predict
+    def split_into_microbatches(self, inputs, targets, microbatch, drop_last):
+        """ Split inputs and targets into microbatch-sized chunks. """
+        # Parse microbatch size
+        if microbatch:
+            if microbatch is True:
+                microbatch = self.microbatch
+            else:
+                microbatch = microbatch or self.microbatch
```
Comment (a-arefina) [inline batchflow/models/torch/base.py:1291]: Since `if microbatch` check above has passed, `or self.microbatch` will never be used

File: batchflow/models/torch/base.py [hunk 10]
```diff
@@ -1167,94 +1151,195 @@ def _train_sam_update_gradients(self, inputs, targets, sync_frequency, sam_rho,
...
+
+        # Compute batch_size and make sure it is the same for all inputs and targets
+        batch_size = len(inputs[0])
+        for i, item in enumerate(inputs):
+            if len(item) != batch_size:
+                raise ValueError('All of `inputs` should have the same length, as the first one!'
+                                    f'Input at position `{i}` has size {len(item)}!={batch_size}')
```
Comment (a-arefina) [inline batchflow/models/torch/base.py:1298]:
```suggestion
                raise ValueError('All of `inputs` should have the same batch size!'
                                    f'Input at position `{i}` has batch size {len(item)}!={batch_size}')
```

File: batchflow/models/torch/base.py [hunk 11]
```diff
@@ -1167,94 +1151,195 @@ def _train_sam_update_gradients(self, inputs, targets, sync_frequency, sam_rho,
...
+        for i, item in enumerate(targets):
+            if len(item) != batch_size:
+                raise ValueError('All of `targets` should have the same length, as the first of `inputs`!'
+                                    f'Target at position `{i}` has size {len(item)}!={batch_size}')
```
Comment (a-arefina) [inline batchflow/models/torch/base.py:1302]:
```suggestion
                raise ValueError('All of `targets` should have the same batch size as `inputs`!'
                                    f'Target at position `{i}` has batch size {len(item)}!={batch_size}')
```

File: batchflow/models/torch/base.py [hunk 12]
```diff
@@ -1167,94 +1151,195 @@ def _train_sam_update_gradients(self, inputs, targets, sync_frequency, sam_rho,
...
+            chunked_targets = [targets]
+
+        return chunked_inputs, chunked_targets, batch_size, microbatch
+
+    def aggregate_microbatches(self, outputs, chunked_outputs, single_output):
+        """ Aggregate outputs from microbatches into outputs for the whole batch.
+        Scalar values are aggregated by `mean`, array values are concatenated along the first (batch) axis.
+        """
+        result = []
+        for i, _ in enumerate(outputs):
```
Comment (a-arefina) [inline batchflow/models/torch/base.py:1325]:
```suggestion
        for i in range(len(outputs)):
```

File: batchflow/models/torch/base.py (hunk 3 above)
Comment (SergeyTsimfer) [reply batchflow/models/torch/base.py:747]: That is the functionality of `ConvBlock`

File: batchflow/models/torch/base.py (hunk 8 above)
Comment (SergeyTsimfer) [reply batchflow/models/torch/base.py:1214]: I really don't like the implementation of it, as it breaks in some cases (although not really relevant here until we try to pass anything but `np.ndarray`s)

Also, it would change one clear pythonic line to just another line of code with a function call, which may do unexpected things (and in the case it indeed would do that)

File: batchflow/models/torch/base.py (hunk 9 above)
Comment (SergeyTsimfer) [reply batchflow/models/torch/base.py:1306]: Nice catch

File: batchflow/models/torch/base.py [hunk 13]
```diff
@@ -117,6 +207,62 @@ def callable_init(module): # example of a callable for init
...
+
+    # Shapes: optional
+    inputs_shapes : sequence
+        Shapes of the input tensors without the batch size.
+        Must be a tuple (one input) or sequence on tuples (multiple inputs) with shapes.
+
+    inputs_shapes : sequence
+        Shapes of the target tensors without the batch size.
+        Must be a tuple (one target) or sequence on tuples (multiple targets) with shapes.
+        Available as `` parameter in the `head` block.
```
[the remark this answers is not in the document]
Comment (HollowPrincess) [reply batchflow/models/torch/base.py:219]: And in other places

File: batchflow/models/torch/base.py [hunk 14]
```diff
@@ -736,44 +729,117 @@ def make_decay(self, decay, optimizer=None, **kwargs):
...
+        After it, create training infrastructure (loss, optimizer, decay).
+
+        The order is defined by `order` key in the config, which is [`initial_block`, `body`, `head`] by default.
+        Each item in `order` should describe the block name, the config name and method to create. It can be a:
+            - string, then we use it as name, config key and method name
+            - tuple of three elements, which are name, config key and method name or callable
+            - dictionary with three items, which are `block_name`, `config_name` and `method`.
 
-    # Define model structure
+            The `block_name` is then used as the identifier in resulting model, i.e. `model.body`, `model.head`.
```
Comment (HollowPrincess) [inline batchflow/models/torch/base.py:752]:
```suggestion
            The `block_name` is used as the identifier in resulting model, i.e. `model.body`, `model.head`.
```
